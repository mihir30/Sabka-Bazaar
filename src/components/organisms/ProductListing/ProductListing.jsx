import React, { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { DropDown, SideBar } from "../../molecules";
import ProductCard from "../../molecules/Product/Product";
import { WithSpinner } from "../../atoms";
import { getData } from "../../utils/common.utils";
import "./ProductListing.scss";

const ProductListing = () => {
  const [productData, setProductData] = useState([]);
  const [isLoading, setIsLoading] = useState(true);
  const [categoryData, setCategoryData] = useState([]);
  const [selectedCatergoryId, setSelectedCatergoryId] = useState();
  const [userSelectedProducts, setUserSelectedProducts] = useState([]);
  let navigate = useNavigate();
  const plpURL = window.location.href;
  let extractCategoryID = plpURL.split("/products/")[1];

  useEffect(() => {
    const fetchData = async () => {
      try {
        const categories = await getData("http://localhost:5000/categories");
        setCategoryData(categories);
        const productsData = await getData("http://localhost:5000/products");
        setProductData(productsData);
        if (!extractCategoryID) {
          setUserSelectedProducts(productsData);
        } else {
          const filterData = productsData.filter(
            (product) => product.category === extractCategoryID
          );
          setUserSelectedProducts(filterData);
        }
        setIsLoading(false);
      } catch (error) {
        console.log("error", error.message);
      }
    };
    fetchData();
  }, [plpURL]);

  const handleCategoryChange = (id) => {
    if (selectedCatergoryId === id) {
      setSelectedCatergoryId(null);
      navigate(`/products/`);
      return setUserSelectedProducts(productData);
    } else {
      const filterData = productData.filter(
        (product) => product.category === id
      );
      setUserSelectedProducts(filterData);
      setSelectedCatergoryId(id);
      navigate(`/products/${id}`);
    }
  };

  const handleDropDownChange = (e) => {
    const id = e.target.value;
    if (id === "Default") {
      navigate(`/products/`);
      return setUserSelectedProducts(productData);
    } else {
      const filterData = productData.filter(
        (product) => product.category === id
      );
      setUserSelectedProducts(filterData);
      navigate(`/products/${id}`);
    }
  };

  return (
    <div>
      {isLoading ? (
        <WithSpinner></WithSpinner>
      ) : (
        <div className="product-listing">
          <SideBar
            categoryData={categoryData}
            id={selectedCatergoryId}
            name="select"
            handleCategoryChange={handleCategoryChange}
          ></SideBar>
          <div className="category-dropdown">
            <DropDown
              categoryData={categoryData}
              handleDropDownChange={handleDropDownChange}
            ></DropDown>
          </div>
          <div className="category-description">
            {userSelectedProducts.map((product) => (
              <ProductCard
                key={product.id}
                id={product.id}
                imageUrl={product.imageURL}
                name={product.name}
                price={product.price}
                stock={product.stock}
                text={product.description}
              ></ProductCard>
            ))}
          </div>
        </div>
      )}
    </div>
  );
};

export default ProductListing;