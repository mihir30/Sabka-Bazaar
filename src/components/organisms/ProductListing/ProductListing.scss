.product-listing {
  .category-container {
    display: none;
  }

  .category-description {
    display: block;
    text-align: center;
    align-items: center;
    margin-top: 35px;

  }

  .category-dropdown {
    display: block;
  }
}

@media (min-width: 768px) {
  .product-listing {
    width: 100%;
    display: flex;

    .category-description {
      display: grid;
      grid-template-columns: 1fr 1fr;
      grid-column-gap: 0px;
      grid-row-gap: 20px;
    }

    .category-dropdown {
      display: none;
    }

    .category-container {
      width: 100%;
      display: block;
    }
  }
}


@media (min-width: 1024px) {
  .product-listing {

    .category-description {
      display: grid;
      grid-template-columns: 1fr 1fr 1fr 1fr;
      grid-row-gap: 20px;
    }
  }
}