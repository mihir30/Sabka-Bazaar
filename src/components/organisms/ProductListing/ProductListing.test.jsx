import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import {render} from '@testing-library/react';
import renderer from 'react-test-renderer';
import ReactDom, { unmountComponentAtNode } from 'react-dom';
import ProductListing from './ProductListing';

describe("Product Listing Component", () => {
    it("should render product listing component", () => {
        const divEl = document.createElement('div');
        ReactDom.render(<BrowserRouter><ProductListing /></BrowserRouter>, divEl);
    });

    let container = null;

    beforeEach(() => {
        // set up a DOM element as a render target
        container = document.createElement("div");
        document.body.appendChild(container);
    });

    afterEach(() => {
        // cleanup on exiting
        unmountComponentAtNode(container);
        container.remove();
        container = null;
    });

    function sleep(ms) {
        return new Promise((resolve) => setTimeout(resolve, ms));
    }

    it("should fetch product data corectly", async () => {

        render(<BrowserRouter><ProductListing /></BrowserRouter>, container);
        await sleep(500);
    });

    it("should have the right productlisting component snapshot", () => {
        const tree = renderer.create(<BrowserRouter><ProductListing /></BrowserRouter>).toJSON();
        expect(tree).toMatchSnapshot();

    });

});