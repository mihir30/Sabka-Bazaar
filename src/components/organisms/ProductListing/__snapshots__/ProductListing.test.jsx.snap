// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Listing Component should have the right productlisting component snapshot 1`] = `
<div>
  <div
    className="spinner-overlay"
  >
    <div
      className="spinner-container"
    />
  </div>
</div>
`;
